test-data/unit/check-narrowing.test

```diff
@@ -1261,3 +1261,30 @@ def g() -> None:
         def foo(): ...
     foo()
 [builtins fixtures/dict.pyi]
+
+
+[case testNarrowingOptionalEqualsNone]
+from typing import Optional
+
+class A: ...
+
+val: Optional[A]
+
+if val == None:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+else:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+if val != None:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+else:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+
+if val in (None,):
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+else:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+if val not in (None,):
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
+else:
+    reveal_type(val)  # N: Revealed type is "Union[__main__.A, None]"
```

Comment:
I'm not sure what you're asking. This test passes.

Mypy doesn't support type narrowing for these cases yet, I'm only fixing its buggy behavior here in this PR.

Reply:
Ahh, got it now! 👌

Reply:
(I assume actually narrowing is much more work?)

Reply:
It's more tricky for sure, and quite rare in Python code. Not as easy as supporting `is None`, since classes (even a subclas of `str`) can override `__eq__` and return `True` when compared to `None`.

I looked into this code because I want to solve https://github.com/python/mypy/issues/9718, I'd rather put my effort there for now :)